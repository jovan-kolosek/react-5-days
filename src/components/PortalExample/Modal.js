import React, { Component } from 'react';
import ReactDOM from 'react-dom';

class Modal extends Component {
  constructor(props) {
    super(props);
    this.modal = document.getElementById('modal-root');
    // Create a div that we'll render the modal into. Because each
    // Modal component has its own element, we can render multiple
    // modal components into the modal container.
    this.el = document.createElement('div');
  }

  componentDidMount() {
    // Append the element into the DOM on mount. We'll render
    // into the modal container element (see the HTML tab).
    this.modal.appendChild(this.el);
  }

  componentWillUnmount() {
    // Remove the element from the DOM when we unmount
    this.modal.removeChild(this.el);
  }

  render() {
    // Use a portal to render the children into the element
    return ReactDOM.createPortal(
      this.props.children,
      this.el,
    );
  }
}

// The Modal component is a normal React component, so we can
// render it wherever we like without needing to know that it's
// implemented with portals.
export default Modal;
