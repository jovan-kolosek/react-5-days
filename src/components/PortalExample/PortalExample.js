import React, { Component } from 'react';
import Modal from './Modal';
import './PortalExample.css';
import RenderContext from '../ContextExample/ContextExample';


class PortalExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = {showModal: false};
  }

  handleShow = () => {
    this.setState({
      showModal: true
    });
  }

  handleHide = () => {
    this.setState({
      showModal: false
    });
  }

  renderModal() {
    if (!this.state.showModal) {
      return null;
    }
    return (
      <Modal>
        <div className="modal">
          <div>
            {
              `With a portal, we can render content into a different
              part of the DOM, as if it were any other React child.`
            }
          </div>
          <button onClick={this.handleHide}>
            {`Hide modal`}
          </button>
        </div>
      </Modal>
    )
  }

  render() {
    return (
      <RenderContext.Consumer>
      {
        toRender => {
          if (toRender === 'portal') {
            return (
              <div className="app">
                { `This div has overflow: hidden.` }
                <button onClick={this.handleShow}>
                  { `Show modal` }
                </button>
                { this.renderModal() }
              </div>
            );
          } else {
            return null;
          }
        }
      }
      </RenderContext.Consumer>
    );
  }
}

export default PortalExample;
