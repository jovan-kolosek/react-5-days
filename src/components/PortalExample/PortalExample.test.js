import React from 'react';
import RenderContext from '../ContextExample/ContextExample';
import PortalExample from './PortalExample';
import renderer from 'react-test-renderer';

test('PortalExample with set context renders', () => {
  const component = renderer.create(
    <RenderContext.Provider value="portal">
      <PortalExample />
    </RenderContext.Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('PortalExample wont render with default context (other than portal)', () => {
  const component = renderer.create(
    <PortalExample />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
