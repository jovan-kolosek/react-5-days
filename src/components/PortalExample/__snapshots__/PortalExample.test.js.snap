// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortalExample with set context renders 1`] = `
<div
  className="app"
>
  This div has overflow: hidden.
  <button
    onClick={[Function]}
  >
    Show modal
  </button>
</div>
`;

exports[`PortalExample wont render with default context (other than portal) 1`] = `null`;
